lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -2,47 +2,217 @@
+        return false
+      end
+
+      plan.model_attrs.each do |model, attrs_set|
+        return false if attrs_set.empty?
+
+        attrs_set.each do |attrs|
+          return false unless RedisMemo::MemoizeQuery
+            .memoized_columns(model)
+            .include?(attrs.keys.sort)
+        end
+      end
+
+      true
+    end
+
+    #
+    # Extracted bind params is hash of sets: each key is a model class, each
+    # value is a set of hashes for memoized column conditions. Example:
+    #
+    #   {
+    #      Site => [
+    #        {name: 'a', city: 'b'},
+    #        {name: 'a', city: 'c'},
+    #        {name: 'b', city: 'b'},
+    #        {name: 'b', city: 'c'},
+    #      ],
+    #   }
+    #
+    def extract!
+      return if operator.nil?
+
+      left.extract!
+      right.extract!
+      __send__(:"#{operator}!")
+    end
+
     def params
-      #
-      # Bind params is hash of sets: each key is a model class, each value is a
-      # set of hashes for memoized column conditions. Example:
-      #
-      #   {
-      #      Site => [
-      #        {name: 'a', city: 'b'},
-      #        {name: 'a', city: 'c'},
-      #        {name: 'b', city: 'b'},
-      #        {name: 'b', city: 'c'},
-      #      ],
-      #   }
-      #
       @params ||= Hash.new do |models, model|
-        models[model] = []
+        models[model] = Set.new
       end
     end
 
-    def union(other)
-      return unless other
+    protected
+
+    # BindParams is built recursively when iterating through the Arel AST
+    # nodes. BindParams represents a binary tree. Query parameters are added to
+    # the leaf nodes of the tree, and the leaf nodes are connected by
+    # operators, such as `union` (or conditions) or `product` (and conditions).
+    attr_accessor :left
+    attr_accessor :right
+    attr_accessor :operator
+
+    # Prior to actually extracting the bind parameters, we first quickly
+    # estimate if it makes sense to do so. If a query contains too many
+    # dependencies, or contains dependencies that have not been memoized, then
+    # the query itself cannot be cached correctly/efficiently, so there’s no
+    # point to actually extract.
+    #
+    # The planning phase is similar to the extraction phase. Though in the
+    # planning phase, we can ignore all the actual attribute values and only
+    # look at the attribute names. This way, we can compute the attribute names
+    # for each set of dependencies without populating their actual values.
+    #
+    # For example, in the planning phase,
+    #
+    #   {a:nil} x {b: nil} => {a: nil, b: nil}
+    #   {a:nil, b:nil} x {a: nil: b: nil} => {a: nil, b: nil}
+    #
+    # and in the extraction phase, that's where the # of dependency can
+    # actually grow significantly:
+    #
+    #   {a: [1,2,3]} x {b: [1,2,3]} => [{a: 1, b: 1}, ....]
+    #   {a:[1,2], b:[1,2]} x {a: [1,2,3]: b: [1,2,3]} => [{a: 1, b: 1}, ...]
+    #
+    class Plan
```

Comment:
nit: Could you move this class to the bottom of the file? Unless there's a good reason to put it here, I found it a bit hard to read since it breaks up the `plan!` and `extract!` methods.

Also maybe put all the attr_accessors of the BindParam class together